### DISCUSSION TITLE
Ubuntu 21.04 + raspberry pi 4: not loading kernel via tftp

### DISCUSSION THREAD

**Discussion by AdlIzzadin** [ltsp/ltsp#455] [Q&A]
Hi,
I installed Ubuntu 21.04 on for small server and use raspberry 4-2gb as thinclient. 
1. RPI4 prepare for netboot.
2. Followed instructions on https://ltsp.org/docs/installation/ and https://ltsp.org/docs/installation/raspios/
RPI4 gets IP 192.168.67.32 but not load kernel via TFTP.
Please help.

**Comment by alkisg** [comment]
Run `journalctl -fu dnsmasq` on the server,
reboot the rpi client,
and upload the server terminal output,
and a photo of the client screen.

**Reply by AdlIzzadin** [reply]
May 03 10:10:23 myi5ws systemd[1]: Starting dnsmasq - A lightweight DHCP and caching DNS server...
May 03 10:10:23 myi5ws dnsmasq[19213]: started, version 2.84rc2 DNS disabled
May 03 10:10:23 myi5ws dnsmasq[19213]: compile time options: IPv6 GNU-getopt DBus no-UBus i18n IDN2 DHCP DHCPv6 no-Lua TFTP conntrack ipset auth cryptohash DNSSEC loop-detect inotify dumpfile
May 03 10:10:23 myi5ws dnsmasq-dhcp[19213]: DHCP, proxy on subnet 192.168.3.0
May 03 10:10:23 myi5ws dnsmasq-dhcp[19213]: DHCP, IP range 192.168.67.20 -- 192.168.67.250, lease time 12h
May 03 10:10:23 myi5ws dnsmasq-tftp[19213]: TFTP root is /srv/tftp
May 03 10:10:23 myi5ws systemd[1]: Started dnsmasq - A lightweight DHCP and caching DNS server.
May 03 10:58:01 myi5ws dnsmasq-dhcp[19213]: DHCPDISCOVER(enp5s1) e4:5f:01:0c:6e:28
May 03 10:58:01 myi5ws dnsmasq-dhcp[19213]: DHCPOFFER(enp5s1) 192.168.67.32 e4:5f:01:0c:6e:28
May 03 10:58:14 myi5ws dnsmasq-dhcp[19213]: DHCPDISCOVER(enp5s1) e4:5f:01:0c:6e:28
May 03 10:58:14 myi5ws dnsmasq-dhcp[19213]: DHCPOFFER(enp5s1) 192.168.67.32 e4:5f:01:0c:6e:28

**Comment by alkisg** [comment]
If that's all the output, then I guess you missed some steps in the raspios page:

https://ltsp.org/docs/installation/raspios/#server-preparation

Try to follow them all. If you ever see any errors in the terminal, stop and mention them.

**Reply by AdlIzzadin** [reply]
I make this instruction without get error:
>install -m 0660 -g sudo /usr/share/ltsp/common/ltsp/ltsp.conf /etc/ltsp/ltsp.conf
>mcedit /etc/ltsp/ltsp.conf -> add in [server] section RPI_IMAGE="raspios"
>ltsp kernel raspios
>ltsp initrd
>ltsp nfs
>echo '/srv/ltsp/raspios  *(rw,async,crossmnt,no_subtree_check,no_root_squash,insecure)' >/etc/exports.d/ltsp-raspios.exports
>exportfs -ra
>echo 'ip=dhcp root=/dev/nfs nfsroot=192.168.67.1:/srv/ltsp/raspios,vers=3,tcp,nolock init=/usr/share/ltsp/client/init/init ltsp.image=images/raspios.img console=serial0,115200 console=tty1 elevator=deadline fsck.repair=yes rootwait quiet splash >plymouth.ignore-serial-consoles modprobe.blacklist=bcm2835_v4l2' >/srv/ltsp/raspios/boot/cmdline.txt
>ltsp image raspios --mksquashfs-params='-comp lzo'

But kernel not loading via TFTP.

**Comment by alkisg** [comment]
What's the output of this command?

ls -l /srv/tftp

**Reply by AdlIzzadin** [reply]
root@myi5ws:/srv/ltsp/raspios# ls -l /srv/tftp
итого 4
lrwxrwxrwx 1 root root   38 May  3 11:15 bcm2708-rpi-b.dtb -> ../ltsp/raspios/boot/bcm2708-rpi-b.dtb
lrwxrwxrwx 1 root root   43 May  3 11:15 bcm2708-rpi-b-plus.dtb -> ../ltsp/raspios/boot/bcm2708-rpi-b-plus.dtb
lrwxrwxrwx 1 root root   43 May  3 11:15 bcm2708-rpi-b-rev1.dtb -> ../ltsp/raspios/boot/bcm2708-rpi-b-rev1.dtb
lrwxrwxrwx 1 root root   39 May  3 11:15 bcm2708-rpi-cm.dtb -> ../ltsp/raspios/boot/bcm2708-rpi-cm.dtb
lrwxrwxrwx 1 root root   41 May  3 11:15 bcm2708-rpi-zero.dtb -> ../ltsp/raspios/boot/bcm2708-rpi-zero.dtb
lrwxrwxrwx 1 root root   43 May  3 11:15 bcm2708-rpi-zero-w.dtb -> ../ltsp/raspios/boot/bcm2708-rpi-zero-w.dtb
lrwxrwxrwx 1 root root   40 May  3 11:15 bcm2709-rpi-2-b.dtb -> ../ltsp/raspios/boot/bcm2709-rpi-2-b.dtb
lrwxrwxrwx 1 root root   40 May  3 11:15 bcm2710-rpi-2-b.dtb -> ../ltsp/raspios/boot/bcm2710-rpi-2-b.dtb
lrwxrwxrwx 1 root root   40 May  3 11:15 bcm2710-rpi-3-b.dtb -> ../ltsp/raspios/boot/bcm2710-rpi-3-b.dtb
lrwxrwxrwx 1 root root   45 May  3 11:15 bcm2710-rpi-3-b-plus.dtb -> ../ltsp/raspios/boot/bcm2710-rpi-3-b-plus.dtb
lrwxrwxrwx 1 root root   40 May  3 11:15 bcm2710-rpi-cm3.dtb -> ../ltsp/raspios/boot/bcm2710-rpi-cm3.dtb
lrwxrwxrwx 1 root root   40 May  3 11:15 bcm2711-rpi-400.dtb -> ../ltsp/raspios/boot/bcm2711-rpi-400.dtb
lrwxrwxrwx 1 root root   40 May  3 11:15 bcm2711-rpi-4-b.dtb -> ../ltsp/raspios/boot/bcm2711-rpi-4-b.dtb
lrwxrwxrwx 1 root root   40 May  3 11:15 bcm2711-rpi-cm4.dtb -> ../ltsp/raspios/boot/bcm2711-rpi-cm4.dtb
lrwxrwxrwx 1 root root   33 May  3 11:15 bootcode.bin -> ../ltsp/raspios/boot/bootcode.bin
lrwxrwxrwx 1 root root   32 May  3 11:15 cmdline.txt -> ../ltsp/raspios/boot/cmdline.txt
lrwxrwxrwx 1 root root   31 May  3 11:15 config.txt -> ../ltsp/raspios/boot/config.txt
lrwxrwxrwx 1 root root   34 May  3 11:15 COPYING.linux -> ../ltsp/raspios/boot/COPYING.linux
lrwxrwxrwx 1 root root   33 May  3 11:15 fixup4cd.dat -> ../ltsp/raspios/boot/fixup4cd.dat
lrwxrwxrwx 1 root root   31 May  3 11:15 fixup4.dat -> ../ltsp/raspios/boot/fixup4.dat
lrwxrwxrwx 1 root root   33 May  3 11:15 fixup4db.dat -> ../ltsp/raspios/boot/fixup4db.dat
lrwxrwxrwx 1 root root   32 May  3 11:15 fixup4x.dat -> ../ltsp/raspios/boot/fixup4x.dat
lrwxrwxrwx 1 root root   33 May  3 11:15 fixup_cd.dat -> ../ltsp/raspios/boot/fixup_cd.dat
lrwxrwxrwx 1 root root   30 May  3 11:15 fixup.dat -> ../ltsp/raspios/boot/fixup.dat
lrwxrwxrwx 1 root root   33 May  3 11:15 fixup_db.dat -> ../ltsp/raspios/boot/fixup_db.dat
lrwxrwxrwx 1 root root   32 May  3 11:15 fixup_x.dat -> ../ltsp/raspios/boot/fixup_x.dat
lrwxrwxrwx 1 root root   30 May  3 11:15 issue.txt -> ../ltsp/raspios/boot/issue.txt
lrwxrwxrwx 1 root root   32 May  3 11:15 kernel7.img -> ../ltsp/raspios/boot/kernel7.img
lrwxrwxrwx 1 root root   33 May  3 11:15 kernel7l.img -> ../ltsp/raspios/boot/kernel7l.img
lrwxrwxrwx 1 root root   32 May  3 11:15 kernel8.img -> ../ltsp/raspios/boot/kernel8.img
lrwxrwxrwx 1 root root   31 May  3 11:15 kernel.img -> ../ltsp/raspios/boot/kernel.img
lrwxrwxrwx 1 root root   37 May  3 11:15 LICENCE.broadcom -> ../ltsp/raspios/boot/LICENCE.broadcom
drwxr-xr-x 3 root root 4096 May  3 11:15 ltsp
lrwxrwxrwx 1 root root   29 May  3 10:32 overlays -> ../ltsp/raspios/boot/overlays
lrwxrwxrwx 1 root root   33 May  3 11:15 start4cd.elf -> ../ltsp/raspios/boot/start4cd.elf
lrwxrwxrwx 1 root root   33 May  3 11:15 start4db.elf -> ../ltsp/raspios/boot/start4db.elf
lrwxrwxrwx 1 root root   31 May  3 11:15 start4.elf -> ../ltsp/raspios/boot/start4.elf
lrwxrwxrwx 1 root root   32 May  3 11:15 start4x.elf -> ../ltsp/raspios/boot/start4x.elf
lrwxrwxrwx 1 root root   33 May  3 11:15 start_cd.elf -> ../ltsp/raspios/boot/start_cd.elf
lrwxrwxrwx 1 root root   33 May  3 11:15 start_db.elf -> ../ltsp/raspios/boot/start_db.elf
lrwxrwxrwx 1 root root   30 May  3 11:15 start.elf -> ../ltsp/raspios/boot/start.elf
lrwxrwxrwx 1 root root   32 May  3 11:15 start_x.elf -> ../ltsp/raspios/boot/start_x.elf

**Comment by alkisg** [comment]
Did you update rpi4's EEPROM code?
https://ltsp.org/docs/installation/raspios/#client-configuration

Also try with a SD card inside the rpi4, that has a fat32 file system with only bootcode.bin in it. You can find it in /srv/ltsp/raspios/boot/bootcode.bin.

**Reply by AdlIzzadin** [reply]
Don't understand. 
I prepared RPI4 for netboot via eeprom update(Instruction give from https://linuxhit.com/raspberry-pi-pxe-boot-netbooting-a-pi-4-without-an-sd-card/), because "raspberry_pi4_network_boot_beta.md" not found on SD card. Now my RPI4 have last offical pieeprom-2021-03-17.bin with netboot option.
![rpi4_boot](https://user-images.githubusercontent.com/72731948/116857299-3c374080-ac05-11eb-884b-2b5cc76e2342.png)

**Comment by alkisg** [comment]
Maybe they changed something in the netbooting code, not sure.
Can you test with bootcode.bin in an sd card?

**Reply by AdlIzzadin** [reply]
Yes. Tested. Nothing changed.
![rpi4_boot1](https://user-images.githubusercontent.com/72731948/116863847-92a97c80-ac0f-11eb-9137-7bb32acec436.png)

**Comment by alkisg** [comment]
From the photo, it seems that it didn't load the firmware from the SD card.
Ask the Raspberry guys on how to load the firmware from the SD card, instead of the one you flashed into the EEPROM.
